pkg/download/download.go

```diff
@@ -55,43 +58,53 @@ func download(client *http.Client, u *url.URL, path string, extract bool) error
 		}
 	}()
 
-	req, err := http.NewRequest(http.MethodGet, u.String(), nil)
-	if err != nil {
-		return err
-	}
-
-	resp, err := client.Do(req)
-	if err != nil {
-		return err
-	}
-	defer resp.Body.Close()
-
-	r := io.Reader(resp.Body)
+	operation := func() error {
+		req, err := http.NewRequest(http.MethodGet, u.String(), nil)
+		if err != nil {
+			return err
+		}
 
-	// extract (optional)
-	if extract {
-		switch {
-		case strings.HasSuffix(u.Path, "gz"):
-			gr, err := gzip.NewReader(resp.Body)
-			if err != nil {
-				return err
-			}
-			r = gr
-		case strings.HasSuffix(u.Path, "bz2"):
-			r = bzip2.NewReader(resp.Body)
-		case strings.HasSuffix(u.Path, "xz"):
-			xzr, err := xz.NewReader(resp.Body)
-			if err != nil {
-				return err
+		resp, err := client.Do(req)
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+
+		r := io.Reader(resp.Body)
+
+		// extract (optional)
+		if extract {
+			switch {
+			case strings.HasSuffix(u.Path, "gz"):
+				gr, err := gzip.NewReader(resp.Body)
+				if err != nil {
+					return err
+				}
+				r = gr
+			case strings.HasSuffix(u.Path, "bz2"):
+				r = bzip2.NewReader(resp.Body)
+			case strings.HasSuffix(u.Path, "xz"):
+				xzr, err := xz.NewReader(resp.Body)
+				if err != nil {
+					return err
+				}
+				r = xzr
+			default:
+				return fmt.Errorf("unknown extension: %s", u.Path)
 			}
-			r = xzr
-		default:
-			return fmt.Errorf("unknown extension: %s", u.Path)
 		}
+
+		if _, err := io.Copy(tmpFile, r); err != nil {
+			return err
```

Comment:
It if retries due to an error here, then `tmpFile` may have been written in the previous iteration?

Reply:
So maybe truncate it at the beginning of `operation`?

Reply:
great catch!  i added file truncation at the beginning of the loop